New tutorial: Partitioned Burgers eq. 1D (#670)
precice/tutorials: partitioned-burgers-1d/README.md

Reviewer: Add references section with your thesis.

{% note %}
The surrogate participant requires PyTorch and related dependencies. By default, the CPU version is installed. If you wish to use the GPU version, see the comment in `neumann-surrogate/requirements.txt`. The GPU version requires several gigabytes of disk space (~6.2Gb).
{% endnote %}

## Running the simulation

### Running the participants

To run the partitioned simulation, open two separate terminals and start each participant individually:

You can find the corresponding `run.sh` script for running the case in the folders corresponding to the participant you want to use:

```bash
cd dirichlet-scipy
./run.sh
```

and

```bash
cd neumann-scipy
./run.sh
```

or, to use the pretrained neural network surrogate participant:

```bash
cd neumann-surrogate
./run.sh
```

### Initial condition

The initial condition file `initial_condition.npz` is automatically generated by the run scripts if it does not exist.
You can also manually generate it using the script in `utils/`:

```bash

> python3 utils/generate_ic.py --epoch <seed>

Reviewer 2: An example here (as in the script) would be useful.

Reviewer 2: Added more details.

```

This script requires the Python libraries `numpy` and `matplotlib`.

### Helper scripts

There are helper scripts in the `utils/` directory that automate runs and visualization of both participants. They also accept an integer seed argument to specify the initial condition.

```bash
./utils/run-partitioned-scipy.sh
```

and

```bash
./utils/run-partitioned-surrogate.sh
```

### Monolithic solution (reference)

You can run the whole domain using the monolithic solver for comparison:

```bash
cd solver-scipy
./run.sh
```

## Post-processing

After both participants (and/or monolithic simulation) have finished, you can run the visualization script.
`visualize_partitioned_domain.py` generates plots comparing the partitioned and monolithic solutions. You can specify which timestep to plot. Call from the root of the tutorial:

```bash

> python3 utils/visualize_partitioned_domain.py --neumann neumann-surrogate/surrogate.npz [timestep]

Reviewer 2: An example would be useful here, I am not sure if I should add an integer here and what good values would be.

Similarly in other places.

Author: Added a comment and changed `[timestep]` to `10`.

You can find the corresponding `run.sh` script for running the case in the folders corresponding to the participant you want to use:

```bash
cd dirichlet-scipy
./run.sh
```

and

```bash
cd neumann-scipy
./run.sh
```

or, to use the pretrained neural network surrogate participant:

```bash
cd neumann-surrogate
./run.sh
```

### Initial condition

The initial condition file `initial_condition.npz` is automatically generated by the run scripts if it does not exist.
You can also manually generate it using the `utils/generate_ic.py` script:

```bash
python3 utils/generate_ic.py
```

This script requires the Python libraries `numpy` and `matplotlib`. It accepts an optional argument `--epoch` as a random number generator seed, which defaults to zero.

To generate the training data, you can use the `utils/generate-training-data.sh` script from the tutorial root directory, which will generate data for different `--epoch` values:

```bash
./utils/generate-training-data.sh 
```

### Monolithic solution (reference)

You can run the whole domain using the monolithic solver for comparison:

```bash
cd solver-scipy
./run.sh
```

## Post-processing

> After both participants (and/or monolithic simulation) have finished, you can run the visualization script.

Reviewer 2: ```suggestion
After both participants have finished, you can run the visualization script.
```

This will not work for the monolithic participant, and the script does not imply so. It will also fail with the error:

```
(venv) tutorials/partitioned-burgers-1d [partitioned-burgers-1d]$ python3 utils/visualize_partitioned_domain.py --neumann solver-scipy/full_domain.npz 10
Loading data from tutorials/partitioned-burgers-1d/utils/../dirichlet-scipy/dirichlet.npz
Loading data from tutorials/partitioned-burgers-1d/utils/../solver-scipy/full_domain.npz
Full domain shape: (101, 384)
Loading Monolithic data from /home/gc/repos/precice/tutorials/partitioned-burgers-1d/utils/../solver-scipy/full_domain.npz
Monolithic shape: (101, 256)
Saved plot to output/full_domain_timestep_slice.png
Traceback (most recent call last):
  File "/home/gc/repos/precice/tutorials/partitioned-burgers-1d/utils/visualize_partitioned_domain.py", line 80, in <module>
    residual = full_solution_history[TIMESTEP_TO_PLOT, :] - solution_gt[TIMESTEP_TO_PLOT, :]
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ValueError: operands could not be broadcast together with shapes (384,) (256,) 

```

The domain is partitioned into participants at $x=1$:

- **Dirichlet**: Solves the left half $[0,1]$ and receives Dirichlet boundary conditions at the interface.
- **Neumann**: Solves the right half $[1,2]$ and receives Neumann boundary conditions at the interface.

Both outer boundaries use zero-gradient conditions $\frac{\partial u}{\partial x} = 0$. The problem can be solved for different initial conditions of superimposed sine waves, which can be generated using the provided script `utils/generate_ic.py`.

<p align="center">
  <img src="images/tutorials-partitioned-burgers-1d-full-domain-diagram.png" alt="Domain Diagram" width="500"/>
  <br><em>Diagram of the partitioned domain with an example initial condition.</em>
</p>

## Configuration

preCICE configuration (image generated using the [precice-config-visualizer](https://precice.org/tooling-config-visualization.html)):

<p align="center">
  <img src="images/tutorials-partitioned-burgers-1d-precice-config.png" alt="preCICE configuration visualization" width="600"/>

> </p>

Reviewer: I don't think we need html for the images in the README, we normally don't use it here.

```suggestion
![preCICE configuration visualization](images/tutorials-partitioned-burgers-1d-precice-config.png)
```

Similarly everywhere here.

Author: I think I wanted the html there to center the images and change their width, but we can do without. Updated.

> ## Available Solvers

Reviewer: ```suggestion
## Available solvers
```

The conservative formulation of the Burgers' equation `solver-scipy` is implemented in a first-order finite volume code using Lax-Friedrichs fluxes and implicit Euler time stepping.

This tutorial includes two versions for the Neumann participant:

- A standard finite volume solver (`neumann-scipy`).

> - A pre-trained neural network surrogate that approximates the solver (`neumann-surrogate`).

Reviewer: ```suggestion
Currently, the SciPy solver can be used for both side, the NN surrogate solver only for the Neumann side.

- SciPy. Simple finite volume solver using Lax-Friedrichs fluxes and implicit Euler time stepping.
- Surrogate. Pre-trained neural network surrogate.
```

To  be consistent with other tutorials.
TODO: It would be nice to write a few more words on the surrogate here.

{% note %}
The surrogate participant requires PyTorch and related dependencies, which requires several gigabytes of disk space (~7Gb).
{% endnote %}

## Running the simulation

### Running the participants

To run the partitioned simulation, open two separate terminals and start each participant individually:

You can find the corresponding `run.sh` script for running the case in the folders corresponding to the participant you want to use:

```bash
cd dirichlet-scipy
./run.sh
```

and

```bash
cd neumann-scipy
./run.sh
```

or, to use the pretrained neural network surrogate participant:

```bash
cd neumann-surrogate
./run.sh
```

### Initial condition

The initial condition file `initial_condition.npz` is automatically generated by the run scripts if it does not exist.
You can also manually generate it using the script in `utils/`:

```bash
python3 utils/generate_ic.py --epoch <seed>
```

This script requires the Python libraries `numpy` and `matplotlib`.

> ---

Reviewer: ```suggestion
```
Not used in the docs, but should also not be necessary here.

### Helper scripts

There are helper scripts in the `utils/` directory that automate runs and visualization of both participants. They also accept an integer seed argument to specify the initial condition.

```bash
./utils/run-partitioned-scipy.sh
```

and

```bash
./utils/run-partitioned-surrogate.sh
```

### Monolithic solution (reference)

You can run the whole domain using the monolithic solver for comparison:

```bash
cd solver-scipy
./run.sh
```

## Visualization

After both participants (and/or monolithic simulation) have finished, you can run the visualization script.

> `visualize_partitioned_domain.py` generates plots comparing the partitioned and monolithic solutions. You can specify which timestep to plot:

Reviewer: ```suggestion
`visualize_partitioned_domain.py` generates plots comparing the partitioned and monolithic solutions. You can specify which timestep to plot. Call from the root of the tutorial:
```

> `visualize_partitioned_domain.py` generates plots comparing the partitioned and monolithic solutions. You can specify which timestep to plot:

Reviewer: How can I compare partitioned and monolithic solutions? If I just run it after one setup, I only get that one setup.

Author: Yes. You have to run both the partitioned and monolithic cases first. The visualize script will draw the monolithic solution conditionally if it exists. For this reason I have the helper scripts to run everything with one command.

Reviewer 2: The documentation currently states

> After both participants (and/or monolithic simulation) have finished, you can run the visualization script.

I think this is enough.

```bash

> python3 visualize_partitioned_domain.py --neumann neumann-surrogate/surrogate.npz [timestep]

Reviewer: ```suggestion
python3 utils/visualize_partitioned_domain.py --neumann neumann-surrogate/surrogate.npz [timestep]
```

```

The script will produce the following output files in the `output/` directory:

> - `full-domain-timestep-slice.png`: Solution $u$ at a selected timestep

Reviewer: ```suggestion
- `full-domain-timestep-slice.png`: Solution $u$ at the selected timestep
```

<p align="left">
  <img src="images/tutorials-partitioned-burgers-1d-full-domain-timestep-slice.png" alt="Full Domain Timestep Slice" width="400"/>
</p>

> - `gradient-timestep-slice.png`: Gradient $du/dx$ at a selected timestep

Reviewer: ```suggestion
- `gradient-timestep-slice.png`: Gradient $du/dx$ at the selected timestep
```

---

> title: Partitioned 1D Burgers' Equation

Reviewer: ```suggestion
title: Partitioned Burgers' equation 1D
```

permalink: tutorials-partitioned-burgers-1d.html
keywords: Python, Neural Network, Surrogate, Burgers Equation, Finite Volume, CFD
summary: This tutorial demonstrates the partitioned solution of the 1D Burgers' equation using preCICE and a neural network surrogate solver.
---

{% note %}
Get the [case files of this tutorial](https://github.com/precice/tutorials/tree/master/partitioned-burgers-1d). Read how in the [tutorials introduction](https://precice.org/tutorials.html).
{% endnote %}

## Setup

We solve the 1D viscous Burgers' equation on the domain $[0,2]$:

$$
\frac{\partial u}{\partial t} = \nu \frac{\partial^2 u}{\partial x^2} - u \frac{\partial u}{\partial x},
$$

where $u(x,t)$ is the scalar velocity field and $\nu$ is the viscosity. In this tutorial by default $\nu$ is very small ($10^{-12}$), but can be changed in the solver.

The domain is partitioned into participants at $x=1$:

- **Dirichlet**: Solves the left half $[0,1]$ and receives Dirichlet boundary conditions at the interface.
- **Neumann**: Solves the right half $[1,2]$ and receives Neumann boundary conditions at the interface.

> Both outer boundaries use zero-gradient conditions $\frac{\partial u}{\partial x} = 0$. The problem is solved for different initial conditions of superimposed sine waves, which can be generated using the provided script.

Reviewer: Which script, how/when do I need to run it?

Author: utils/generate_ic.py

<p align="center">
  <img src="images/tutorials-partitioned-burgers-1d-initial-condition.png" alt="Initial Condition" width="400"/>
  <br><em>Initial condition used for the simulation (seed 0, see <code>generate_ic.py</code>)</em>
</p>

The conservative formulation of the Burgers' equation `solver-scipy` is implemented in a first-order finite volume code using Lax-Friedrichs fluxes and implicit Euler time stepping.

This tutorial includes two versions for the Neumann participant:
- A standard finite volume solver (`neumann-scipy`).

> - A pre-trained neural network surrogate that approximates the solver (`neumann-surrogate`).

Reviewer: This information goes into subsection "Available solvers"

## Configuration

preCICE configuration (image generated using the [precice-config-visualizer](https://precice.org/tooling-config-visualization.html)):

<p align="center">
  <img src="images/tutorials-partitioned-burgers-1d-precice-config.png" alt="preCICE configuration visualization" width="600"/>
</p>

> ## Running the tutorial

Reviewer: ```suggestion
## Running the simulation
```

> ### Initial condition 

Reviewer: needed for training, or also for running?
It could be easier if we only explain the "running" here and hide the training part in `neumann-surrogate`?

Author: I don't have the train scripts here (yet)

Reviewer 2: Linked to https://github.com/vidulejs/neural-adapter, that's enough for now.

Before running the participants, you must generate the initial condition file. From the root of the tutorial folder, run the command and replace `<seed>` with any integer.:

```bash
python3 generate_ic.py --epoch <seed>

> ```

Reviewer: Is this really a script that is general to all solvers? (otherwise, put in subfolders)
still, a `tools` subfolder could be good`

This script requires the Python libraries `numpy` and `matplotlib`. It generates the initial condition file `initial_condition.npz` used by both participants. If you do not have the dependencies, then you can run either of the participant run scripts to create a Python virtual environment with the required packages and try again. 

Or, simply, run the helper scripts (see below), which will also generate the initial condition file if it does not exist.

---

#### Helper scripts

There are helper scripts that automate runs and visualization of both participants. They also accept an integer seed argument to specify the initial condition.

```bash
run-partitioned-scipy.sh
```

and

```bash
run-partitioned-surrogate.sh

> ```

Reviewer: We normally don't provide such global running scripts. Envision a future state, where we include this case into the [system tests](https://precice.org/dev-docs-system-tests.html).

Author: I put these in the new utils folder. There are other scripts there as well.

### Running the participants

To run the partitioned simulation, open two separate terminals and start each participant individually:

You can find the corresponding `run.sh` script for running the case in the folders corresponding to the participant you want to use:

```bash
cd dirichlet-scipy
./run.sh
```

and

```bash
cd neumann-scipy
./run.sh
```

or, to use the pretrained neural network surrogate participant:

```bash
cd neumann-surrogate
./run.sh
```

> **Note:** The surrogate participant requires PyTorch and related dependencies, which requires several gigabytes of disk space.

Reviewer: For notes, use the notes environment: https://precice.org/docs-meta-cheatsheet.html#alerts

### Monolithic solution (reference)

You can run the whole domain using the monolithic solver for comparison:

```bash
cd solver-scipy
./run.sh
```

 
## Visualization

After both participants (and/or monolithic simulation) have finished, you can run the visualization script.
`visualize_partitioned_domain.py` generates plots comparing the partitioned and monolithic solutions. You can specify which timestep to plot:

```bash
python3 visualize_partitioned_domain.py --neumann neumann-surrogate/surrogate.npz [timestep]
```

> The script will produce the following output files in the `images/` directory:

Reviewer: Maybe not a good idea to also put this into `images/`, which contains the images for the documentation/website.

---
title: Partitioned Burgers' equation 1D
permalink: tutorials-partitioned-burgers-1d.html
keywords: Python, Neural Network, Surrogate, Burgers Equation, Finite Volume, CFD
summary: This tutorial demonstrates the partitioned solution of the 1D Burgers' equation using preCICE and a neural network surrogate solver.
---

{% note %}

> Get the [case files of this tutorial](https://github.com/precice/tutorials/tree/master/partitioned-burgers-1d). Read how in the [tutorials introduction](https://precice.org/tutorials.html).

Reviewer 2: Updated format: https://github.com/precice/tutorials/pull/704
